File: buyv_flutter_app/cors_proxy_server.js
const express = require('express');
const cors = require('cors');
const https = require('https');
const { createProxyMiddleware } = require('http-proxy-middleware');

const app = express();
const PORT = 3001;

// Allowed origins including local previews
const allowedOrigins = [
  'http://localhost:5500',
  'http://127.0.0.1:5500',
  'http://localhost:52000',
  'http://localhost:3000',
  'http://localhost:8080'
];

const corsOptions = {
  origin: function (origin, callback) {
    // allow non-browser requests (no origin) and whitelisted origins
    if (!origin || allowedOrigins.includes(origin)) {
      callback(null, true);
    } else {
      callback(new Error('Not allowed by CORS'));
    }
  },
  methods: ['GET', 'POST', 'PUT', 'DELETE', 'OPTIONS'],
  allowedHeaders: ['Content-Type', 'Authorization', 'CJ-Access-Token'],
  credentials: false // do not use wildcard with credentials=true
};

// Enable CORS globally
app.use(cors(corsOptions));

// Parse JSON bodies
app.use(express.json());

// Preflight handler for CJ routes (must be BEFORE proxy)
app.options('/api/cj/*', (req, res) => {
  const origin = req.headers.origin;
  if (!origin || allowedOrigins.includes(origin)) {
    res.header('Access-Control-Allow-Origin', origin || '*');
    res.header('Vary', 'Origin');
    res.header('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
    res.header('Access-Control-Allow-Headers', 'Content-Type, Authorization, CJ-Access-Token');
    res.header('Access-Control-Max-Age', '86400');
    return res.sendStatus(200);
  }
  return res.sendStatus(403);
});

// Proxy middleware for CJ Dropshipping API
const cjProxy = createProxyMiddleware({
  target: 'https://developers.cjdropshipping.com',
  changeOrigin: true,
  secure: true,
  xfwd: true,
  timeout: 30000,
  proxyTimeout: 30000,
  agent: new https.Agent({ keepAlive: true }),
  headers: {
    // mimic a real browser user agent to avoid upstream filtering issues
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120 Safari/537.36'
  },
  pathRewrite: {
    '^/api/cj': '/api2.0/v1'
  },
  onProxyReq: (proxyReq, req, res) => {
    console.log(`Proxying ${req.method} ${req.url} to CJ API`);
    // Forward CJ-Access-Token if present
    // Forward CJ-Access-Token if present
    if (req.headers['cj-access-token']) {
      proxyReq.setHeader('CJ-Access-Token', req.headers['cj-access-token']);
    }
  },
  onProxyRes: (proxyRes, req, res) => {
    console.log(`Response from CJ API: ${proxyRes.statusCode}`);
    const origin = req.headers.origin;
    // Add CORS headers to response
    proxyRes.headers['Access-Control-Allow-Origin'] = origin || '*';
    proxyRes.headers['Vary'] = 'Origin';
    proxyRes.headers['Access-Control-Allow-Methods'] = 'GET, POST, PUT, DELETE, OPTIONS';
    proxyRes.headers['Access-Control-Allow-Headers'] = 'Content-Type, Authorization, CJ-Access-Token';
    proxyRes.headers['Access-Control-Expose-Headers'] = 'Content-Type, Authorization, CJ-Access-Token';
  },
  onError: (err, req, res) => {
    console.error('Proxy error:', err);
    res.status(500).json({
      error: 'Proxy error',
      message: err.message
    });
  }
});

// Use the proxy for all /api/cj routes
app.use('/api/cj', cjProxy);

// Health check endpoint
app.get('/health', (req, res) => {
  res.json({ 
    status: 'OK', 
    message: 'CJ Dropshipping CORS Proxy Server is running',
    timestamp: new Date().toISOString()
  });
});

// Handle preflight requests
app.options('*', (req, res) => {
  res.header('Access-Control-Allow-Origin', '*');
  res.header('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
  res.header('Access-Control-Allow-Headers', 'Content-Type, Authorization, CJ-Access-Token');
  res.sendStatus(200);
});

app.listen(PORT, () => {
  console.log(`🚀 CJ Dropshipping CORS Proxy Server running on http://localhost:${PORT}`);
  console.log(`📡 Proxying requests to: https://developers.cjdropshipping.com`);
  console.log(`🔗 Use this base URL in your Flutter app: http://localhost:${PORT}/api/cj`);
  console.log(`💡 Health check: http://localhost:${PORT}/health`);
});

// Graceful shutdown
process.on('SIGTERM', () => {
  console.log('🛑 CORS Proxy Server shutting down...');
  process.exit(0);
});

process.on('SIGINT', () => {
  console.log('🛑 CORS Proxy Server shutting down...');
  process.exit(0);
});

File: buyv_flutter_app/cors_proxy_server.dart
import 'dart:io';
import 'dart:convert';

/// CORS Proxy Server for CJ Dropshipping API
/// This server acts as a proxy to bypass CORS restrictions when calling CJ API from Flutter web
class CORSProxyServer {
  static const int port = 3001;
  static const String cjBaseUrl = 'https://developers.cjdropshipping.com/api2.0/v1';
  
  static HttpServer? _server;
  static final HttpClient _httpClient = HttpClient();

  /// Start the CORS proxy server
  static Future<void> start() async {
    try {
      _server = await HttpServer.bind(InternetAddress.anyIPv4, port);
      print('🚀 CORS Proxy Server started on http://localhost:$port');
      print('📡 Proxying requests to: $cjBaseUrl');
      print('🔄 Ready to handle CJ Dropshipping API requests...\n');

      await for (HttpRequest request in _server!) {
        await _handleRequest(request);
      }
    } catch (e) {
      print('❌ Failed to start CORS proxy server: $e');
    }
  }

  /// Handle incoming HTTP requests
  static Future<void> _handleRequest(HttpRequest request) async {
    try {
      // Add CORS headers to all responses
      _addCORSHeaders(request.response);

      // Handle preflight OPTIONS requests
      if (request.method == 'OPTIONS') {
        request.response.statusCode = HttpStatus.ok;
        await request.response.close();
        return;
      }

      // Extract the CJ API path from the request
      String path = request.uri.path;
      if (path.startsWith('/api/cj')) {
        path = path.substring('/api/cj'.length);
      }

      // Build the target URL
      final targetUrl = '$cjBaseUrl$path';
      final queryString = request.uri.query.isNotEmpty ? '?${request.uri.query}' : '';
      final fullUrl = '$targetUrl$queryString';

      print('🔄 Proxying ${request.method} request to: $fullUrl');

      // Create the proxy request
      final proxyRequest = await _httpClient.openUrl(request.method, Uri.parse(fullUrl));

      // Copy headers from original request (except host and origin)
      request.headers.forEach((name, values) {
        if (!_shouldSkipHeader(name)) {
          for (String value in values) {
            proxyRequest.headers.add(name, value);
          }
        }
      });

      // Set proper host header for CJ API
      proxyRequest.headers.set('host', 'developers.cjdropshipping.com');
      proxyRequest.headers.set('origin', 'https://developers.cjdropshipping.com');
      proxyRequest.headers.set('referer', 'https://developers.cjdropshipping.com/');

      // Copy request body if present
      HttpClientResponse proxyResponse;
      if (request.contentLength > 0) {
        // Read request body and write to proxy request
        final requestBody = await request.toList();
        for (var chunk in requestBody) {
          proxyRequest.add(chunk);
        }
        proxyResponse = await proxyRequest.close();
      } else {
        proxyResponse = await proxyRequest.close();
      }

      // Copy response status and headers
      request.response.statusCode = proxyResponse.statusCode;
      
      proxyResponse.headers.forEach((name, values) {
        if (!_shouldSkipResponseHeader(name)) {
          for (String value in values) {
            request.response.headers.add(name, value);
          }
        }
      });

      // Ensure CORS headers are present
      _addCORSHeaders(request.response);

      // Copy response body
      await proxyResponse.pipe(request.response);

      print('✅ Request completed with status: ${proxyResponse.statusCode}');

    } catch (e) {
      print('❌ Error handling request: $e');
      
      // Send error response
      request.response.statusCode = HttpStatus.internalServerError;
      _addCORSHeaders(request.response);
      
      final errorResponse = {
        'error': 'Proxy server error',
        'message': e.toString(),
        'timestamp': DateTime.now().toIso8601String(),
      };
      
      request.response.headers.contentType = ContentType.json;
      request.response.write(jsonEncode(errorResponse));
      await request.response.close();
    }
  }

  /// Add CORS headers to the response
  static void _addCORSHeaders(HttpResponse response) {
    response.headers.set('Access-Control-Allow-Origin', '*');
    response.headers.set('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
    response.headers.set('Access-Control-Allow-Headers', 
        'Origin, X-Requested-With, Content-Type, Accept, Authorization, CJ-Access-Token');
    response.headers.set('Access-Control-Max-Age', '86400');
    response.headers.set('Access-Control-Allow-Credentials', 'false');
  }

  /// Check if a request header should be skipped when proxying
  static bool _shouldSkipHeader(String headerName) {
    final skipHeaders = {
      'host',
      'origin',
      'referer',
      'connection',
      'upgrade-insecure-requests',
      'sec-fetch-site',
      'sec-fetch-mode',
      'sec-fetch-dest',
      'sec-ch-ua',
      'sec-ch-ua-mobile',
      'sec-ch-ua-platform',
    };
    return skipHeaders.contains(headerName.toLowerCase());
  }

  /// Check if a response header should be skipped when returning to client
  static bool _shouldSkipResponseHeader(String headerName) {
    final skipHeaders = {
      'access-control-allow-origin',
      'access-control-allow-methods',
      'access-control-allow-headers',
      'access-control-max-age',
      'access-control-allow-credentials',
    };
    return skipHeaders.contains(headerName.toLowerCase());
  }

  /// Stop the proxy server
  static Future<void> stop() async {
    if (_server != null) {
      await _server!.close();
      _server = null;
      print('🛑 CORS Proxy Server stopped');
    }
    _httpClient.close();
  }

  /// Get server status
  static bool get isRunning => _server != null;
  
  /// Get server URL
  static String get serverUrl => 'http://localhost:$port';
}

/// Main function to run the CORS proxy server
void main() async {
  print('🌐 Starting CORS Proxy Server for CJ Dropshipping API...');
  print('📋 This server will help bypass CORS restrictions for Flutter web app');
  print('🔧 Server configuration:');
  print('   - Local port: ${CORSProxyServer.port}');
  print('   - Target API: ${CORSProxyServer.cjBaseUrl}');
  print('   - CORS enabled for all origins');
  print('');

  // Handle Ctrl+C gracefully
  ProcessSignal.sigint.watch().listen((signal) async {
    print('\n🛑 Received interrupt signal, shutting down...');
    await CORSProxyServer.stop();
    exit(0);
  });

  try {
    await CORSProxyServer.start();
  } catch (e) {
    print('❌ Failed to start server: $e');
    exit(1);
  }
}